apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/service/blockchain/TransactionServiceImpl.java

```diff
@@ -0,0 +1,280 @@
+        this.toEntityConverter = toEntityConverter;
+    }
+
+    @Override
+    public Transaction findTransaction(long transactionId) {
+        TransactionEntity entity = transactionDao.findTransaction(transactionId, databaseManager.getDataSource());
+        return toModelConverter.convert(entity);
+    }
+
+    @Override
+    public Transaction findTransaction(long transactionId, int height) {
+        TransactionalDataSource dataSource = getDatasourceWithShardingByTransactionId(transactionId);
+        TransactionEntity entity = transactionDao.findTransaction(transactionId, height, dataSource);
+        return toModelConverter.convert(entity);
+    }
+
+    @Override
+    public Transaction findTransactionByFullHash(byte[] fullHash) {
+        TransactionEntity entity = transactionDao.findTransactionByFullHash(fullHash, databaseManager.getDataSource());
+        return toModelConverter.convert(entity);
+    }
+
+    @Override
+    public Transaction findTransactionByFullHash(byte[] fullHash, int height) {
+        TransactionalDataSource dataSource = getDatasourceWithShardingByTransactionId(Convert.fullHashToId(fullHash));
+        TransactionEntity entity = transactionDao.findTransactionByFullHash(fullHash, height, dataSource);
+        return toModelConverter.convert(entity);
+    }
+
+    @Override
+    public boolean hasTransaction(long transactionId) {
+        return transactionDao.hasTransaction(transactionId, databaseManager.getDataSource());
+    }
+
+    @Override
+    public boolean hasTransaction(long transactionId, int height) {
+        return transactionDao.hasTransaction(transactionId, height, databaseManager.getDataSource());
+    }
+
+    @Override
+    public boolean hasTransactionByFullHash(byte[] fullHash) {
+        return transactionDao.hasTransactionByFullHash(fullHash, databaseManager.getDataSource());
+    }
+
+    @Override
+    public boolean hasTransactionByFullHash(byte[] fullHash, int height) {
+        return transactionDao.hasTransactionByFullHash(fullHash, height, databaseManager.getDataSource());
+    }
+
+    @Override
+    public byte[] getFullHash(long transactionId) {
+        return transactionDao.getFullHash(transactionId, databaseManager.getDataSource());
+    }
+
+    @Override
+    public List<Transaction> findBlockTransactions(long blockId) {
+        TransactionalDataSource dataSource = blockIndexService.getDataSourceWithSharding(blockId);
+        List<TransactionEntity> transactions = transactionDao.findBlockTransactions(blockId, dataSource);
+        return transactions.stream().map(toModelConverter).collect(Collectors.toList());
+    }
+
+    @Override
+    public long getBlockTransactionsCount(long blockId) {
+        TransactionalDataSource dataSource = blockIndexService.getDataSourceWithSharding(blockId);
+        return transactionDao.getBlockTransactionsCount(blockId, dataSource);
+    }
+
+    @Override
+    public void saveTransactions(List<Transaction> transactions) {
```

Comment:
Null check and 'zero collection size' check would be helpful here. Do nothing for empty collection.

Reply:
done;